### PR TITLE
Require Zendesk Code.org users to have oauth logins

### PR DESCRIPTION
**Pull request by bethanyaconnor**
[Jira](https://codedotorg.atlassian.net/browse/FND-707)

I had some trouble testing this other than unit tests, so let me know if there's a better way to test locally.

### CODE REVIEW

File: dashboard/app/controllers/zendesk_session_controller.rb
Comment (islemaster) [inline dashboard/app/controllers/zendesk_session_controller.rb]: This would be a good spot for a comment explaining _why_ we restrict Zendesk SSO for code.org accounts that don't have matching Google OAuth configured.

File: dashboard/app/models/user.rb
Comment (islemaster) [inline dashboard/app/models/user.rb]: I don't think this line makes sense.  `google_oauth` doesn't appear to be a method on `User`, nor does it make sense to compare the boolean result of an `include?` check with the user's email address.

At this point, can we simply drop support for unmigrated accounts in this check?

File: dashboard/app/models/user.rb
Comment (islemaster) [inline dashboard/app/models/user.rb]: I think this is both more terse and more flexible in the case that an account has multiple linked Google accounts (a possibility in our data model, although our UI doesn't encourage it.)
```suggestion
      authentication_options.any? {|ao| ao.google_oauth? && ao.email == email}
```

File: dashboard/app/models/user.rb
Comment (islemaster) [inline dashboard/app/models/user.rb]: It's not clear to me from this method name what the expected behavior is when the user doesn't have a Google account linked.  Maybe `has_google_oauth_matching_primary_email?` for our use case?

**Review by islemaster — COMMENTED** [review]
Is it sufficient to verify that the user has a linked, matching Google account, or do we also need to check that the user last signed in through that account? (As opposed to signing in with a linked password, or another provider.)

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: Oh yeah that's way better!

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: Changed! This method went through a couple of iterations so the name wasn't well chosen.

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: Interestingly, there are 10 accounts with a @code.org email that are unmigrated and only 2 of them belong to people who no longer work here:

```
select email from users where email like "%@code.org" and provider != "migrated";
+------------------------+
| email                  |
+------------------------+
| karis.kim@code.org     |
| josh.schulte1@code.org |
| ha@code.org            |
| drew.perlman@code.org  |
| maddie.kasula@code.org |
| eric@code.org          |
| eric@code.org          |
| kristina@code.org      |
| winter@code.org        |
| will@code.org          |
+------------------------+

```

File: dashboard/app/models/user.rb
Comment (sureshc) [inline dashboard/app/models/user.rb]: I don't believe I put this requirement in this Jira Task, but could we make this more restrictive and only allow a single Authentication Option (`google_oauth`)?  That is a requirement of some of the related Tasks in this Epic.

Going further down this line of reasoning, could we consolidate some of the logic in the ZenDesk SSO controller into this method and rename it to `employee?` (or something similar like `internal_cdo?`).  The method would require:

- User email domain = code.org
- only authentication option is Google auth
- User email = authentication option email

The method could also require/assume the user is migrated, since we have more flexibility to impose that on internal Code.org users.

Comment (islemaster) [reply dashboard/app/models/user.rb]: :+1: An `employee?` check sounds like a great idea.  I think it might need to be even more strict than you describe though, if we want to potentially exclude facilitators and/or contributors we've granted `code.org` addresses to.

Comment (sureshc) [reply dashboard/app/models/user.rb]: If we can't easily distinguish between employees and contributors when someone authenticates on Code Studio using Google OAuth, then a method that identifies the broader set of internal code.org users would suffice, for now (`internal_cdo?`).

Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: My concern with making google SSO the only option is that I don't know of a good way to remove a password if one is associated with an account (I couldn't create a query that finished in any reasonable amount of time to actually find out if this is an issue).

There are a handful of code.org employees but I do feel better about telling them to migrate, though I don't know why they aren't. I thought we migrated over most users...

I'll refactor the functions today!

Comment (sureshc) [reply dashboard/app/models/user.rb]: I pulled a report of all employee logins and whether they have Google OAuth enabled.  I'll need to work with all of the people on that list to update their accounts before we merge this functionality.  I'll sync up with you on that this afternoon.

**Review by sureshc — APPROVED** [review]
I haven't finished getting everyone to migrate their Code Studio accounts to using Google authentication, so please hold off on merging until I've done that.

**Comment by bethanyaconnor** [comment]
Thanks! Will wait for the go-ahead from you to merge.

**Pull request closed without merging**
